### PR TITLE
Feature/users

### PR DESCRIPTION
**Pull request by R1ZZU**

### CODE REVIEW

File: app/models/User.coffee [hunk 1]
```diff
@@ -0,0 +1,21 @@
+co       = require 'co'
+mongoose = require 'mongoose'
+
+userSchema = new mongoose.Schema
+	displayName:
+		type: String
+		required: yes
+	id:
+		type: String
+		unique: true
```
Comment (DaQuirm) [inline app/models/User.coffee]: also `yes`?

File: app/models/User.coffee [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+		required: yes
+
+userSchema.static 'fromOAuthProfile', co.wrap (profile) ->
+	id = "#{profile.provider}#{profile.id}"
+
+	found = yield @findOne(id: id).exec()
+	if found then found else @create
```
Comment (DaQuirm) [inline app/models/User.coffee]: Does this look better?
```coffeescript
found or @create ...
```

File: app/models/User.coffee (hunk 1 above)
Comment (R1ZZU) [inline app/models/User.coffee]: yes!

File: app/models/User.coffee (hunk 2 above)
Comment (R1ZZU) [inline app/models/User.coffee]: :hand: :neutral_face:

File: package.json [hunk 3]
```diff
@@ -53,14 +54,16 @@
     "webpack-notifier": "^1.2.1"
   },
   "dependencies": {
+    "co": "^4.6.0",
```
Comment (DaQuirm) [inline package.json]: Could you also look into replacing `vo` with `co` when launching Nightmare?

File: test/models/User.spec.coffee [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+		beforeEach ->
+			profile =
+				id: 'id'
+				provider: 'provider'
+				username: 'username'
+
+		afterEach (done) ->
+			User.remove {}, -> done()
+
+		it "should create new user if user with the same id doesn't exist", ->
```
Comment (DaQuirm) [inline test/models/User.spec.coffee]: unnecessary double quotes :cat:

File: test/models/User.spec.coffee [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+			yield User.fromOAuthProfile profile
+			yield checkUserExists 'providerid'
+
+		it 'should return existing user', ->
```
Comment (DaQuirm) [inline test/models/User.spec.coffee]: why write `should`? :pouting_cat:

File: test/models/User.spec.coffee [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+			user.id.should.equal 'providerid'
+
+		it 'should assign profile username to user displayName', ->
+			user = yield User.fromOAuthProfile profile
+			user.displayName.should.equal profile.username
+
+		it 'should assign profile username to user displayName even if profile has displayName', ->
+			profile.displayName = 'displayName'
+			user = yield User.fromOAuthProfile profile
+			user.displayName.should.equal profile.username
```
Comment (DaQuirm) [inline test/models/User.spec.coffee]: And maybe check that it's actually not `'displayName'`

File: test/models/User.spec.coffee [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+
+		it 'should return existing user', ->
+			newUser      = yield User.fromOAuthProfile profile
+			existingUser = yield User.fromOAuthProfile profile
+
+			newUser._id.should.deep.equal existingUser._id
+
+		it 'should join provider name with profile id to create userid', ->
+			user = yield User.fromOAuthProfile profile
+			user.id.should.equal 'providerid'
```
Comment (DaQuirm) [inline test/models/User.spec.coffee]: Replace constant with expression

File: config/var/dev.cson [hunk 8]
```diff
@@ -1,5 +1,7 @@
 {
-	mongo_connection_string: "mongodb://localhost/cssqd-dev"
```
Comment (DaQuirm) [inline config/var/dev.cson]: How come this works with no changes to `index.coffee` or config?? :confused:

File: test/models/User.spec.coffee (hunk 6 above)
Comment (R1ZZU) [inline test/models/User.spec.coffee]: what for?

File: package.json [hunk 9]
```diff
@@ -53,14 +54,16 @@
...
     "cson": "^3.0.1",
+    "cssqd-shared": "file:./shared",
     "koa": "^1.1.0",
     "koa-bodyparser": "^2.0.1",
     "koa-passport": "^1.2.0",
     "koa-router": "^5.2.3",
     "koa-session": "^3.3.1",
     "koa-static": "^1.5.1",
-    "mongoose": "^4.2.3",
+    "mongoose": "^4.2.7",
```
Comment (DaQuirm) [inline package.json]: Newest mongoose, awesome :new:

File: test/models/User.spec.coffee (hunk 6 above)
Comment (DaQuirm) [inline test/models/User.spec.coffee]: Well, `profile.username` is not guaranteed to be different from `'displayName'`
But that's trifle anyway

File: config/var/dev.cson (hunk 8 above)
Comment (R1ZZU) [inline config/var/dev.cson]: wip

File: test/models/User.spec.coffee [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+	describe '#fromOAuthProfile', ->
+
+		beforeEach ->
+			profile =
+				id: 'id'
+				provider: 'provider'
+				username: 'username'
+
+		afterEach (done) ->
+			User.remove {}, -> done()
```
Comment (DaQuirm) [inline test/models/User.spec.coffee]: Why not pass `done`? :smile_cat:

File: config/var/dev.cson (hunk 8 above)
Comment (DaQuirm) [inline config/var/dev.cson]: sure, gj :smile_cat:

File: test/models/User.spec.coffee (hunk 10 above)
Comment (R1ZZU) [inline test/models/User.spec.coffee]: :hand: :neutral_face: (2)

File: test/models/User.spec.coffee (hunk 4 above)
Comment (R1ZZU) [inline test/models/User.spec.coffee]:
```sh
npm run coffeelint
...
✓ ./test/models/User.spec.coffee
...
```

Comment (DaQuirm) [inline test/models/User.spec.coffee]: right, it's because of the apostrophe

Comment (R1ZZU) [inline test/models/User.spec.coffee]: :+1:

File: package.json (hunk 3 above)
Comment (R1ZZU) [inline package.json]: yes, sure

File: test/models/User.spec.coffee (hunk 5 above)
Comment (R1ZZU) [inline test/models/User.spec.coffee]: have a lot of arguments, but none of them is compelling :crying_cat_face:

Comment (R1ZZU) [inline test/models/User.spec.coffee]: https://mochajs.org/example/tests.html
we can ask TJ why :smile:

File: package.json (hunk 3 above)
Comment (R1ZZU) [inline package.json]: let's defer this
https://github.com/DaQuirm/css-quickdraw-redux/issues/48

File: app/models/User.coffee [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+	id:
+		type: String
+		unique: yes
+		required: yes
+	role:
+		type: String
+		enum: ['game_master', 'player']
+		default: 'player'
+
+userSchema.static 'fromOAuthProfile', co.wrap (profile) ->
```
Comment (kirilknysh) [inline app/models/User.coffee]: Does it make sense to create a separate layer for it (smth like DAL/service/controller) to manipulate with DB?

**Comment by kirilknysh** [comment]
Looks good!

**Pull request merged**